lms/services/canvas_files.py

```diff
@@ -0,0 +1,30 @@
+from lms.models import CanvasFile
+
+
+class CanvasFilesService:
+    """Methods for working with the models.CanvasFile DB model."""
+
+    def __init__(self, db):
+        self._db = db
+
+    def upsert(self, canvas_file):
+        existing_file = (
+            self._db.query(CanvasFile)
+            .filter_by(
+                consumer_key=canvas_file.consumer_key,
+                tool_consumer_instance_guid=canvas_file.tool_consumer_instance_guid,
+                course_id=canvas_file.course_id,
+                file_id=canvas_file.file_id,
+            )
+            .one_or_none()
+        )
+
+        if existing_file:
+            existing_file.filename = canvas_file.filename
+            existing_file.size = canvas_file.size
+        else:
+            self._db.add(canvas_file)
```

Comment:
This is a single-method service for now but in https://github.com/hypothesis/lms/pull/2328 it'll grow additional methods for searching for and getting files from the DB

Reply:
Note on the design of this `upsert()` method:

## Fields as separate params :negative_squared_cross_mark: 

In the past we've often written service methods that encapsulate the model classes and just accept the fields as separate arguments, like this:

```python
def upsert(self, consumer_key, tool_consumer_instance_guid, course_id, file_id, filename, size):
    ...
```

This doesn't work well because it creates methods with long argument lists. PyLint rightly complains. Adding a new field to the model class necessitates adding another argument to the method and then updating every place that calls the method (including every test).

## Combine params into value object :negative_squared_cross_mark: 

You can beat the too-many-arguments problem by providing a value class to hold them all in a single object. You could put the value class in `canvas_files.py` or even inside the `CanvasFilesService` class as an inner class:

```python
class CanvasFilesService:

    class CanvasFile(NamedTuple):
        consumer_key: str
        tool_consumer_instance_guid: str
        course_id: int
        file_id: int
        filename: str
        size: int

    def upsert(self, canvas_file):
        ...
```

The view code would then do something like:

```python
canvas_files_svc.upsert(CanvasFilesService.CanvasFile(...))
```

The problem with this is that the `CanvasFilesService.CanvasFile` is pretty much a duplicate of `models.CanvasFile` and we'll end up needing these value classes for many of our model classes

## Combine params into dict :negative_squared_cross_mark: 

Defeat the too-many-arguments problem by putting them all in a data dict:

```python
def upsert(self, data):
    existing_file = (
        self._db.query(CanvasFile)
        .filter_by(
            consumer_key=data["consumer_key"],
            tool_consumer_instance_guid=data["tool_consumer_instance_guid"],
            course_id=data["course_id"],
            file_id=data["file_id"],
        )
        .one_or_none()
    )

    ...
```

But data dicts suck because they just obfuscate the method's actual params. The method effectively does still have a long argument list but you now have to read the entire method body to find out what all the arguments are. This approach also tends to lead to methods that pass their data dicts into other methods wholesale, which is even worse.

## Just use the model! :heavy_check_mark: 

Services don't have to encapsulate models. It's okay for views to instantiate model objects and pass them into services. This is the solution I went for:

```python
def upsert(self, canvas_file):
    ...
```

The view then does:

```python
canvas_files_svc.upsert(models.CanvasFile(consumer_key="foo", ...))
```

This works particularly well with our thin models approach because:

* All model objects have the same interface when it comes to constructing them. You can always pass all the fields as kwargs:

  ```python
  canvas_file = CanvasFile(consumer_key="foo", ...)
  ```

  You can also always assign all the fields:

  ```python
  canvas_file = CanvasFile()
  canvas_file.consumer_key = "foo"
  ...
  ```

  This is the normal behaviour of SQLAlchemy ORM model classes and of Factory Boy factories, and our non-SQLAlchemy models also behave the same.

* You can create a model object without adding it to the DB. You can create a model object that will _never_ get added to the DB and just pass it around as a value. `canvas_file = CanvasFile()` doesn't add anything to the DB session.

* You can create an invalid model object. For example you can create a `CanvasFile` whose `consumer_key` doesn't match any `ApplicationInstance` in the DB, or one that has no `tool_consumer_instance_guid` even though that column isn't nullable, etc. This is totally fine as long as you're just using it as a value and not adding it to the DB (or as long as someone fills out the missing fields before it does get added to the DB).

* Because of our thin models approach (no `__init__()` methods, and few or no methods in general on model classes) our models are close to being value objects (like `NamedTuple`'s) and we can get away with treating them as such